packages/teams-js/src/internal/nestedAppAuth.ts

```diff
@@ -0,0 +1,186 @@
+
+  const extendedWindow = window as unknown as NestedAuthExtendedWindow;
+  if (extendedWindow.nestedAppAuthBridge) {
+    logger('nestedAppAuthBridge already exists on current window, skipping polyfill');
+    return;
+  }
+
+  const nestedAppAuthBridge = createNestedAppAuthBridge(extendedWindow, handlers);
+  if (nestedAppAuthBridge) {
+    extendedWindow.nestedAppAuthBridge = nestedAppAuthBridge;
+  }
+}
+
+const createNestedAppAuthBridgeLogger = nestedAppAuthLogger.extend('createNestedAppAuthBridge');
+
+/**
+ * @hidden
+ * Creates a bridge for nested app authentication.
+ *
+ * @internal
+ * Limited to Microsoft-internal use
+ *
+ * @param {Window | null} window - The window object where the nested app authentication bridge will be created. If null, the function will log an error message and return null.
+ * @returns {NestedAppAuthBridge | null} Returns an object with methods for adding and removing event listeners, and posting messages. If the provided window is null, returns null.
+ *
+ * @property {Function} addEventListener - Adds an event listener to the window. Only supports the 'message' event. If an unsupported event is passed, logs an error message.
+ * @property {Function} postMessage - Posts a message to the window. The message should be a stringified JSON object with a messageType of 'NestedAppAuthRequest'. If the message does not meet these criteria, logs an error message.
+ * @property {Function} removeEventListener - Removes an event listener from the window.
+ */
+function createNestedAppAuthBridge(
+  window: Window | null,
+  bridgeHandlers: NestedAppAuthBridgeHandlers,
+): NestedAppAuthBridge | null {
+  const logger = createNestedAppAuthBridgeLogger;
+
+  if (!window) {
+    logger('nestedAppAuthBridge cannot be created as current window does not exist');
+    return null;
+  }
+
+  const { onMessage, handlePostMessage } = bridgeHandlers;
+  const nestedAppAuthBridgeHandler = (callback: (response: string) => void) => (evt: MessageEvent) =>
+    onMessage(evt, callback);
+
+  return {
+    addEventListener: (eventName, callback): void => {
+      if (eventName === 'message') {
+        window.addEventListener(eventName, nestedAppAuthBridgeHandler(callback));
+      } else {
+        logger(`Event ${eventName} is not supported by nestedAppAuthBridge`);
+      }
+    },
+    postMessage: (message: string): void => {
+      // Validate that it is a valid auth bridge request message
+      const parsedMessage = (() => {
+        try {
+          return JSON.parse(message);
+        } catch (e) {
+          return null;
+        }
+      })();
+
+      if (
+        !parsedMessage ||
+        typeof parsedMessage !== 'object' ||
+        parsedMessage.messageType !== NestedAppAuthMessageEventNames.Request
+      ) {
+        logger('Unrecognized data format received by app, message being ignored. Message: %o', message);
+        return;
+      }
+
+      // Post the message to the top window
+      handlePostMessage(message);
```

Comment:
Security question: How are we protecting ourselves against the possibility of a malicious child iframe sending code to trigger XSS attack?

It's crucial to verify the origin of the received messages to prevent potential cross-site scripting attacks. I'm not sure if the plan is to inject the bridge into trusted child iframes only; even in that case, we still need to consider the possibility of a compromised child iframe. Additionally, if we don't know the child iframe in advance, there might be security implications. This question can be asked in security review.

Reply:
Thanks. It makes sense.